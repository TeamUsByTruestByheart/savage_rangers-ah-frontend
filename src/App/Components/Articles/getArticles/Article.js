import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import propTypes from 'prop-types';
import Rater from 'react-rater';
import ReactHtmlParser from 'react-html-parser';
import ReactImageFallback from 'react-image-fallback';
import { Link } from 'react-router-dom';
import countRating from '../../../../Helpers/countRating';

export class Article extends Component {
  render() {
    const { data } = this.props;
    const {
      body, title, coverImage, createdAt, slug
    } = data;
    const { name } = data.Category;
    const { profileImage, firstName, username } = data.User;
    const { comments, reads, shares } = data.stats;
    const { allUsers, statistics } = data.rating;
    const ratings = countRating(statistics);

    const date = moment(createdAt).format('MMM Do YY');
    const img = 'https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg';
    const defaultProfileImage = 'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png';
    const articleBody = body.length > 200 ? `${body.substring(0, 210)} ...` : body;
    const articleTitle = title.length > 52 ? `${title.substring(0, 52)}...` : title;
    const categoryColor = (name === 'LOVE' && '#ff1744') || (name === 'TECHNOLOGY' && '#3f51b5') || (name === 'MUSIC' && '#880e4f') || (name === 'BUSINESS' && '#2196f3') || (name === 'ART' && '#9c27b0') || '#afad99';
    const names = (firstName === null && username)
      || (firstName === undefined && username)
      || firstName;
    return (
      <React.Fragment>
        <div className="articles container row">
          <div className="articles__cover col-lg-3 col-md-12 col-sm-12">
            <ReactImageFallback
              src={coverImage}
              fallbackImage={img}
              initialImage="loader.gif"
              alt="cool image should be here"
              className="coverImage"
            />
          </div>
          <div className="articles__details col-lg-9 col-md-12 col-sm-12">
            <div className="articles__details--content">
              <Link className="link" to={`/articles/${slug}`}>
                <h3>
                  {articleTitle}
                </h3>
              </Link>
              <div className="category row">
                <div className="col-lg-6 col-md-6 col-sm-12 row">
                  <div className="color" style={{ background: `${categoryColor}` }} />
                  <div className="name col ">
                    <h5>{name}</h5>
                  </div>

                </div>
                <div className="Rate col-lg-6 col-md-6 col-sm-12 row">
                  <div className="rating col">
                    <Rater total={5} rating={ratings} interactive={false} />
                  </div>
                  <div className="ratedUser col-1">
                    <h6>
                      [
                      {allUsers}
                      ]
                    </h6>
                  </div>

                </div>
              </div>
              <p className="articleBody">
                {ReactHtmlParser(articleBody)}
              </p>
            </div>
            <div className="articles__details--Author row">
              <div className="col-lg-2 col-md-6 col-sm-6">
                <ReactImageFallback
                  src={profileImage}
                  fallbackImage={defaultProfileImage}
                  initialImage="loader.gif"
                  alt="cool image should be here"
                  className="proImage"
                />
              </div>
              <div className="profileDetail col-lg-3 col-md-6 col-sm-6">
                <Link className="userProfile link" to={`/profile/${username}`}>
                  <p className="pro">
                    {names}
                    {' '}
                    {data.User.lastName}
                  </p>
                </Link>
                <p>{date}</p>
              </div>
              <div className="articlesBtn col-lg-6 col-md-6 col-sm-12">
                <div className="row">
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="fab fa-readme" />
                      {' '}
                      {''}
                      {reads}
                    </button>
                  </div>
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="far fa-comment-alt" />
                      {' '}
                      {''}
                      {comments}
                    </button>
                  </div>
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="fas fa-share-alt" />
                      {' '}
                      {''}
                      {shares}
                    </button>
                  </div>
                </div>

                <div className="more">
                  <Link className="read-more fa link" to={`/articles/${slug}`}><p>Read more</p></Link>
                </div>
              </div>

            </div>
          </div>

        </div>
      </React.Fragment>
    );
  }
}

Article.propTypes = { data: propTypes.object };

export default connect(state => ({ stats: state.articleData }))(Article);
