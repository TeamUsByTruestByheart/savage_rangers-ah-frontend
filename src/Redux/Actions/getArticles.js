import async from 'async';
import axios from '../../configs/axios';
import { GET_ARTICLES, ARTICLES_ERROR } from './GET_ARTICLES';

export default {
  getArticles: paginate => (dispatch) => {
    axios
      .get(`/api/articles?offset=${paginate}&limit=10`)
      .then((res) => {
        const paginationDetails = res.data.result.pagedArticles;
        const articles = res.data.result.Articles;
        async.each(articles, async (article, callback) => {
          await axios
            .get(`/api/articles/${article.slug}/stats`)
            .then((response) => {
              articles[articles.indexOf(article)].stats = response.data.article.stats;
            });
          await axios
            .get(`/api/articles/${article.slug}/ratings/statistics`)
            .then((response) => {
              articles[articles.indexOf(article)].rating = response.data.data;
            })
            .catch(() => {
              articles[articles.indexOf(article)].rating = { allUsers: 0, statistics: [] };
            });
          callback();
        }, () => {
          articles.pagination = paginationDetails;
          dispatch({ type: GET_ARTICLES, articlesData: articles });
        });
      })
      .catch((error) => {
        dispatch({ type: ARTICLES_ERROR, error });
      });
  }
};
