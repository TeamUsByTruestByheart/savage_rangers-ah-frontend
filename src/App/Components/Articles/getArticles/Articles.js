import React, { Component } from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import Pagination from 'react-pagination-library';
import searchAction from '../../../../Redux/Actions/searchArticle';
import Article from './Article';
import Navbar from '../../Common/NavProfile/navbar';
import Footer from '../../Common/Footer';
import articleActions from '../../../../Redux/Actions/getArticles';
import Loader from '../../Common/loader';

const { getArticles } = articleActions;
const { searchArticles } = searchAction;

let paginate = 1;
export class Articles extends Component {
  state = {
    isLoading: true,
    currentPage: 1,
    articles: [],
    input: '',
    filter: 'title'
  }


  componentWillMount() {
    const { getArticles: getArticlesData } = this.props;
    getArticlesData(paginate);
  }

  componentWillReceiveProps({ articles }) {
    if (articles) {
      setTimeout(() => {
        this.setState({ isLoading: false });
      }, 2000);
    }
  }

  changeCurrentPage = (numPage) => {
    this.setState({ isLoading: true });
    const page = numPage - 1;
    paginate = page > 0 ? `${page}0` : page;
    const { getArticles: getArticlesData } = this.props;
    getArticlesData(Number(paginate));
    this.setState({ currentPage: numPage });
  }

  handleChange = (update) => {
    const newState = update.target.value;
    this.setState({ [update.target.name]: newState });
  }

  handleSearch = () => {
    this.setState({ isLoading: true });
    const { searchArticles: searchData } = this.props;
    const { filter, input } = this.state;
    searchData(input, filter);
  }

  // handleBack = () => {
  //   const { getArticles: getArticlesData } = this.props;
  //   getArticlesData(paginate);
  //   this.forceUpdate();
  // }

  render() {
    const { Search, articles, searchFailed } = this.props;
    const data = Search === undefined ? articles : Search;
    const page = Search === undefined ? 'page' : 'pagination-Container';
    const { isLoading } = this.state;
    let articlesToDisplay;
    if (data) {
      articlesToDisplay = data.map(article => (
        <Article
          key={article.id}
          data={article}
        />
      ));
    }

    let activePage;
    let numberOfPage;
    if (articles) {
      activePage = articles.pagination.currentPage;
      numberOfPage = articles.pagination.pages;
    }
    if (!isLoading) {
      return (
        <div className="main">
          <Navbar />
          <div>
            <div className="row main__search_label">
              <div className="col-lg-3 col-md-3 col-sm-11">
                <div className="page__label" id="label">
                  <p className="page__label--text">Articles</p>
                </div>
              </div>
              <div className="main__search_label--search col-lg-9 col-md-11 col-sm-11 ">
                <div className="main__search-field input-group md-form form-sm form-2 pl-0">
                  <input
                    className="form-control my-0 py-1 red-border"
                    type="text"
                    name="input"
                    value={this.state.input}
                    onChange={this.handleChange}
                    placeholder="Search"
                    aria-label="Search"
                  />
                  <div className="dropdown">
                    <form>
                      <select name="filter" className="custom-select" onChange={({ target }) => this.setState({ filter: target.value })} value={this.state.filter}>
                        <option selected>Search by</option>
                        <option value="tags">Tag</option>
                        <option value="title">Title</option>
                        <option value="username">Author</option>
                        <option value="body">Content</option>
                      </select>
                    </form>
                  </div>

                  <div className="main__search--btn-search input-group-append">
                    <button type="submit" onClick={this.handleSearch}>
                      <span className="searchIcon input-group-text red lighten-3" id="basic-text1">
                        <i
                          className="fas fa-search text-grey"
                          aria-hidden="true"
                        />
                      </span>
                    </button>
                  </div>

                </div>
              </div>
            </div>

            {searchFailed ? (
              <div className="errorPage-articles">
                <div className="errorPage-articles__error-articles">
                  <h1>404</h1>
                  <p>
                    Article not found
                  </p>
                  <a href="/articles" className="back"><h6>Back to Articles</h6></a>
                  {/* <button type="button" onClick={this.setState({ currentPage: 0 })}>Go Back</button> */}
                </div>
              </div>
            )
              : (
                <div>

                  <div className="main__display-articles container">
                    {
                      articlesToDisplay
                    }
                  </div>
                  <div className={page}>
                    <Pagination
                      currentPage={activePage}
                      totalPages={numberOfPage}
                      changeCurrentPage={this.changeCurrentPage}
                    />
                  </div>
                </div>
              )}
            <Footer />
          </div>
        </div>

      );
    }
    return <Loader />;
  }
}

Articles.propTypes = {
  getArticles: propTypes.func.isRequired,
  searchArticles: propTypes.func.isRequired,
  articles: propTypes.array,
  Search: propTypes.array,
  searchFailed: propTypes.object
};


export const mapStateToProps = state => ({
  articles: state.articles.articles,
  Search: state.searchArticle.articles,
  searchFailed: state.searchArticle.searchFailed
});
export default connect(mapStateToProps, { getArticles, searchArticles })(Articles);
