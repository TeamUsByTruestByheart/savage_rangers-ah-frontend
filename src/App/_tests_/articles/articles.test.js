import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import configureStore from 'redux-mock-store';
import { shallow } from '../../../enzyme';
import { Articles, mapStateToProps } from '../../Components/Articles/getArticles/Articles';
import article from './articles';

mapStateToProps({ article: { data: { article: {} } } });
const middlewares = [thunk, promiseMiddleware];
const mockStore = configureStore(middlewares);

describe('<articles />', () => {
  describe('render()', () => {
    it('it should render all articles', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: article.mockArticle }
      };
      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const wrapper = shallow(<Articles store={store} {...props} data />);
      expect(wrapper).toMatchSnapshot();
      expect(store.getState).toMatchSnapshot();
      expect(wrapper.find('div').exists()).toBe(true);
      expect(wrapper.find('navbar').exists()).toBe(true);
      expect(wrapper.find('Footer').exists()).toBe(true);
      expect(typeof article).toBe('object');
    });
  });
});
