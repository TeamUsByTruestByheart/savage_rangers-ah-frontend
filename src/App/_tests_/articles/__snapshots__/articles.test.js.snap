// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<articles /> render() it should render all articles 1`] = `ShallowWrapper {}`;

exports[`<articles /> render() it should render all articles 2`] = `[Function]`;
